packages/kbn-dev-utils/src/kbn_client/kbn_client_requester.ts

```diff
@@ -0,0 +1,119 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Url from 'url';
+
+import Axios from 'axios';
+
+import { isAxiosRequestError, isConcliftOnGetError } from './errors';
+import { ToolingLog } from '../tooling_log';
+
+export const uriencode = (
+  strings: TemplateStringsArray,
+  ...values: Array<string | number | boolean>
+) => {
+  const queue = strings.slice();
+
+  if (queue.length === 0) {
+    throw new Error('how could strings passed to `uriencode` template tag be empty?');
+  }
+
+  if (queue.length !== values.length + 1) {
+    throw new Error('strings and values passed to `uriencode` template tag are unbalanced');
+  }
+
+  let url = queue.shift()!;
+  while (values.length) {
+    url += encodeURIComponent(values.shift()!);
+    url += queue.shift()!;
+  }
+  return url;
+};
+
+const MAX_ATTEMPTS = 5;
+
+export interface ReqOptions {
+  description?: string;
+  path: string;
+  query?: Record<string, any>;
+  method: 'GET' | 'POST' | 'PUT' | 'DELETE';
+  body?: any;
+  attempt?: number;
+}
+
+const delay = (ms: number) =>
+  new Promise(resolve => {
+    setTimeout(resolve, ms);
+  });
+
+export class KbnClientRequester {
+  constructor(private readonly log: ToolingLog, private readonly kibanaUrls: string[]) {}
+
+  private pickUrl() {
+    const url = this.kibanaUrls.shift()!;
+    this.kibanaUrls.push(url);
+    return url;
+  }
+
+  public resolveUrl(relativeUrl: string = '/') {
+    return Url.resolve(this.pickUrl(), relativeUrl);
+  }
+
+  async request<T>(options: ReqOptions): Promise<T> {
+    const url = Url.resolve(this.pickUrl(), options.path);
+    const description = options.description || `${options.method} ${url}`;
+    const attempt = options.attempt === undefined ? 1 : options.attempt;
+
+    try {
+      const response = await Axios.request<T>({
+        method: options.method,
+        url,
+        data: options.body,
+        params: options.query,
+        headers: {
+          'kbn-xsrf': 'kbn-client',
+        },
+      });
+
+      return response.data;
+    } catch (error) {
+      let retryErrorMsg: string | undefined;
+      if (isAxiosRequestError(error)) {
+        retryErrorMsg = `[${description}] request failed (attempt=${attempt})`;
+      } else if (isConcliftOnGetError(error)) {
+        retryErrorMsg = `Conflict on GET (path=${options.path}, attempt=${attempt})`;
+      }
+
+      if (retryErrorMsg) {
+        if (attempt < MAX_ATTEMPTS) {
+          this.log.error(retryErrorMsg);
+          await delay(1000 * attempt);
+          return await this.request<T>({
```

Comment:
Are there any tests or anything where an error state is desired, and so this will now always try 5 times?

Reply:
In this case, no, but I do think it would make sense to expose maxAttempts as an option just in case.